File: Tools/WistoriaDeveloperToolkit.py
import unreal

MAP_PATH = "/Game/Maps/Lvl_CombatTest"
PROJECT_NAME = "MyProject"

ROOT_FOLDERS = [
    "/Game/Maps",
    "/Game/Blueprints",
    "/Game/Characters",
    "/Game/Enemies",
    "/Game/Bosses",
    "/Game/UI",
    "/Game/FX",
    "/Game/Audio",
    "/Game/Materials",
    "/Game/Items",
    "/Game/Quests",
    "/Game/DataAssets",
    "/Game/Dungeons",
    "/Game/Companions",
]

BLUEPRINTS = [
    ("BP_WistoriaGameMode", "/Game/Blueprints", f"/Script/{PROJECT_NAME}.WistoriaGameMode"),
    ("BP_WistoriaPlayerController", "/Game/Blueprints", f"/Script/{PROJECT_NAME}.WistoriaPlayerController"),
    ("BP_WistoriaCharacter", "/Game/Characters", f"/Script/{PROJECT_NAME}.WistoriaCharacter"),
    ("BP_WistoriaEnemy", "/Game/Enemies", f"/Script/{PROJECT_NAME}.WistoriaEnemy"),
    ("BP_DungeonBoss", "/Game/Bosses", f"/Script/{PROJECT_NAME}.WistoriaBossEnemy"),
    ("BP_DungeonZone", "/Game/Dungeons", f"/Script/{PROJECT_NAME}.DungeonZone"),
    ("BP_DungeonWaveSpawner", "/Game/Dungeons", f"/Script/{PROJECT_NAME}.DungeonWaveSpawner"),
    ("BP_SavePoint", "/Game/Blueprints", f"/Script/{PROJECT_NAME}.SavePoint"),
    ("BP_TreasureChest", "/Game/Items", f"/Script/{PROJECT_NAME}.TreasureChest"),
    ("BP_WorldLootPickup", "/Game/Items", f"/Script/{PROJECT_NAME}.WorldLootPickup"),
    ("BP_CompanionMage", "/Game/Companions", f"/Script/{PROJECT_NAME}.WistoriaCompanionCharacter"),
    ("BP_NPCProfessor", "/Game/Blueprints", f"/Script/{PROJECT_NAME}.DialogueNPC"),
    ("BP_FastTravelGate", "/Game/Blueprints", f"/Script/{PROJECT_NAME}.FastTravelGate"),
]

def log(msg):
    unreal.log(f"[Wistoria Toolkit] {msg}")

def ensure_folders():
    for folder in ROOT_FOLDERS:
        if not unreal.EditorAssetLibrary.does_directory_exist(folder):
            unreal.EditorAssetLibrary.make_directory(folder)
            log(f"Created folder: {folder}")

def load_class(path):
    cls = unreal.load_class(None, path)
    if not cls:
        log(f"Missing class: {path}")
    return cls

def create_blueprint(asset_name, path, parent_class_path):
    asset_path = f"{path}/{asset_name}"

    if unreal.EditorAssetLibrary.does_asset_exist(asset_path):
        log(f"Exists: {asset_path}")
        return unreal.load_asset(asset_path)

    parent_class = load_class(parent_class_path)
    if not parent_class:
        return None

    factory = unreal.BlueprintFactory()
    factory.set_editor_property("ParentClass", parent_class)

    asset = unreal.AssetToolsHelpers.get_asset_tools().create_asset(
        asset_name,
        path,
        unreal.Blueprint,
        factory
    )

    if asset:
        log(f"Created Blueprint: {asset_path}")

    return asset

def build_core_blueprints():
    for name, path, parent in BLUEPRINTS:
        create_blueprint(name, path, parent)

def load_bp_class(asset_path):
    asset = unreal.load_asset(asset_path)
    if not asset:
        log(f"Missing asset: {asset_path}")
        return None

    generated_class = asset.generated_class()
    if not generated_class:
        log(f"No generated class for: {asset_path}")
        return None

    return generated_class

def create_combat_test_level():
    world = unreal.EditorLevelLibrary.new_level(MAP_PATH)

    if not world:
        log("Failed to create/open combat test level.")
        return

    log(f"Created/opened level: {MAP_PATH}")

    player_start_class = unreal.load_class(None, "/Script/Engine.PlayerStart")
    navmesh_class = unreal.load_class(None, "/Script/NavigationSystem.NavMeshBoundsVolume")

    enemy_class = load_bp_class("/Game/Enemies/BP_WistoriaEnemy")
    save_point_class = load_bp_class("/Game/Blueprints/BP_SavePoint")
    chest_class = load_bp_class("/Game/Items/BP_TreasureChest")

    unreal.EditorLevelLibrary.spawn_actor_from_class(
        player_start_class,
        unreal.Vector(0, 0, 120),
        unreal.Rotator(0, 0, 0)
    )

    if enemy_class:
        unreal.EditorLevelLibrary.spawn_actor_from_class(
            enemy_class,
            unreal.Vector(600, 0, 120),
            unreal.Rotator(0, 180, 0)
        )

    if save_point_class:
        unreal.EditorLevelLibrary.spawn_actor_from_class(
            save_point_class,
            unreal.Vector(-350, 250, 100),
            unreal.Rotator(0, 0, 0)
        )

    if chest_class:
        unreal.EditorLevelLibrary.spawn_actor_from_class(
            chest_class,
            unreal.Vector(350, 250, 100),
            unreal.Rotator(0, 0, 0)
        )

    nav = unreal.EditorLevelLibrary.spawn_actor_from_class(
        navmesh_class,
        unreal.Vector(0, 0, 0),
        unreal.Rotator(0, 0, 0)
    )

    if nav:
        nav.set_actor_scale3d(unreal.Vector(20, 20, 3))

    unreal.EditorLevelLibrary.save_current_level()
    log("Combat test level built and saved.")

def save_all():
    unreal.EditorAssetLibrary.save_directory("/Game", only_if_is_dirty=True, recursive=True)

def run():
    log("Starting Wistoria Developer Toolkit...")
    ensure_folders()
    build_core_blueprints()
    save_all()
    create_combat_test_level()
    save_all()
    log("Toolkit complete.")

run()

File: Tools/WistoriaDeveloperToolkit_v1_Real/WistoriaDeveloperToolkit/Source/WistoriaDeveloperToolkit/Private/WistoriaDeveloperToolkit.cpp
#include "WistoriaDeveloperToolkit.h"

#include "AssetRegistry/AssetRegistryModule.h"
#include "AssetToolsModule.h"
#include "Editor.h"
#include "Editor/EditorEngine.h"
#include "Factories/BlueprintFactory.h"
#include "FileHelpers.h"
#include "Framework/MultiBox/MultiBoxBuilder.h"
#include "GameFramework/PlayerStart.h"
#include "HAL/FileManager.h"
#include "IAssetTools.h"
#include "LevelEditor.h"
#include "Misc/MessageDialog.h"
#include "Misc/Paths.h"
#include "Modules/ModuleManager.h"
#include "NavigationSystem.h"
#include "NavMesh/NavMeshBoundsVolume.h"
#include "ToolMenus.h"
#include "Engine/DirectionalLight.h"
#include "Engine/StaticMesh.h"
#include "Engine/StaticMeshActor.h"
#include "Engine/SkyLight.h"
#include "UObject/ConstructorHelpers.h"

#define LOCTEXT_NAMESPACE "FWistoriaDeveloperToolkitModule"

void FWistoriaDeveloperToolkitModule::StartupModule()
{
    UToolMenus::RegisterStartupCallback(
        FSimpleMulticastDelegate::FDelegate::CreateRaw(this, &FWistoriaDeveloperToolkitModule::RegisterMenus));
}

void FWistoriaDeveloperToolkitModule::ShutdownModule()
{
    if (UToolMenus::IsToolMenuUIEnabled())
    {
        UnregisterMenus();
    }

    UToolMenus::UnRegisterStartupCallback(this);
}

void FWistoriaDeveloperToolkitModule::RegisterMenus()
{
    UToolMenu* ToolsMenu = UToolMenus::Get()->ExtendMenu("LevelEditor.MainMenu.Tools");
    if (!ToolsMenu)
    {
        return;
    }

    FToolMenuSection& Section = ToolsMenu->AddSection("WistoriaDeveloperToolkit", LOCTEXT("WistoriaSection", "Wistoria"));

    Section.AddMenuEntry(
        "WistoriaValidateProject",
        LOCTEXT("ValidateProject", "Wistoria: Validate Project"),
        LOCTEXT("ValidateProjectTooltip", "Check the project for key Wand and Wistoria folders and classes."),
        FSlateIcon(),
        FUIAction(FExecuteAction::CreateRaw(this, &FWistoriaDeveloperToolkitModule::ValidateProject))
    );

    Section.AddMenuEntry(
        "WistoriaCreateFolders",
        LOCTEXT("CreateFolders", "Wistoria: Create Folder Structure"),
        LOCTEXT("CreateFoldersTooltip", "Create the core /Game folders for Wand and Wistoria."),
        FSlateIcon(),
        FUIAction(FExecuteAction::CreateRaw(this, &FWistoriaDeveloperToolkitModule::CreateFolderStructure))
    );

    Section.AddMenuEntry(
        "WistoriaBuildBlueprints",
        LOCTEXT("BuildCoreBlueprints", "Wistoria: Build Core Blueprints"),
        LOCTEXT("BuildCoreBlueprintsTooltip", "Create Blueprint children from the core C++ gameplay classes."),
        FSlateIcon(),
        FUIAction(FExecuteAction::CreateRaw(this, &FWistoriaDeveloperToolkitModule::BuildCoreBlueprints))
    );

    Section.AddMenuEntry(
        "WistoriaGenerateArena",
        LOCTEXT("GenerateArena", "Wistoria: Generate Combat Test Arena"),
        LOCTEXT("GenerateArenaTooltip", "Populate the current level with a basic arena, lighting, navigation volume, PlayerStart, and enemy."),
        FSlateIcon(),
        FUIAction(FExecuteAction::CreateRaw(this, &FWistoriaDeveloperToolkitModule::GenerateCombatTestArena))
    );

    Section.AddMenuEntry(
        "WistoriaFixRedirectors",
        LOCTEXT("FixRedirectors", "Wistoria: Fix Redirectors"),
        LOCTEXT("FixRedirectorsTooltip", "Save all assets and ask the Content Browser to clean redirectors manually if required."),
        FSlateIcon(),
        FUIAction(FExecuteAction::CreateRaw(this, &FWistoriaDeveloperToolkitModule::FixRedirectors))
    );
}

void FWistoriaDeveloperToolkitModule::UnregisterMenus()
{
    UToolMenus::UnregisterOwner(this);
}

void FWistoriaDeveloperToolkitModule::ShowMessage(const FText& Title, const FText& Message) const
{
    FMessageDialog::Open(EAppMsgType::Ok, Message, Title);
}

void FWistoriaDeveloperToolkitModule::ValidateProject()
{
    const TArray<FString> RequiredFolders = {
        TEXT("/Game/Maps"),
        TEXT("/Game/Blueprints"),
        TEXT("/Game/Characters"),
        TEXT("/Game/Enemies"),
        TEXT("/Game/Bosses"),
        TEXT("/Game/UI"),
        TEXT("/Game/Items"),
        TEXT("/Game/Quests"),
        TEXT("/Game/DataAssets"),
        TEXT("/Game/Dungeons"),
        TEXT("/Game/Companions")
    };

    FString Report;
    int32 MissingCount = 0;

    for (const FString& Folder : RequiredFolders)
    {
        if (FPackageName::DoesPackageExist(Folder))
        {
            Report += FString::Printf(TEXT("OK: %s\n"), *Folder);
        }
        else
        {
            MissingCount++;
            Report += FString::Printf(TEXT("Missing: %s\n"), *Folder);
        }
    }

    const bool bHasCharacterClass = FindObject<UClass>(ANY_PACKAGE, TEXT("WistoriaCharacter")) != nullptr || LoadObject<UClass>(nullptr, TEXT("/Script/MyProject.WistoriaCharacter")) != nullptr;
    Report += bHasCharacterClass ? TEXT("OK: WistoriaCharacter C++ class\n") : TEXT("Missing: WistoriaCharacter C++ class\n");

    const FText Title = LOCTEXT("ValidationTitle", "Wistoria Project Validation");
    const FText Message = FText::FromString(FString::Printf(TEXT("Missing folders: %d\n\n%s"), MissingCount, *Report));
    ShowMessage(Title, Message);
}

void FWistoriaDeveloperToolkitModule::CreateFolderStructure()
{
    const TArray<FString> Folders = {
        TEXT("Maps"), TEXT("Blueprints"), TEXT("Characters"), TEXT("Enemies"), TEXT("Bosses"),
        TEXT("UI"), TEXT("FX"), TEXT("Audio"), TEXT("Materials"), TEXT("Items"),
        TEXT("Quests"), TEXT("DataAssets"), TEXT("Dungeons"), TEXT("Companions")
    };

    const FString ContentDir = FPaths::ProjectContentDir();

    for (const FString& Folder : Folders)
    {
        IFileManager::Get().MakeDirectory(*(ContentDir / Folder), true);
    }

    FAssetRegistryModule::AssetCreated(nullptr);
    ShowMessage(LOCTEXT("FoldersCreatedTitle", "Wistoria Toolkit"), LOCTEXT("FoldersCreatedMessage", "Core Wand and Wistoria folders have been created under /Game."));
}

bool FWistoriaDeveloperToolkitModule::CreateBlueprintAsset(const FString& AssetName, const FString& PackagePath, const FString& ParentClassPath)
{
    const FString FullAssetPath = PackagePath / AssetName;
    if (LoadObject<UObject>(nullptr, *FullAssetPath))
    {
        UE_LOG(LogTemp, Log, TEXT("Wistoria Toolkit: Blueprint already exists: %s"), *FullAssetPath);
        return true;
    }

    UClass* ParentClass = LoadObject<UClass>(nullptr, *ParentClassPath);
    if (!ParentClass)
    {
        UE_LOG(LogTemp, Warning, TEXT("Wistoria Toolkit: Missing parent class: %s"), *ParentClassPath);
        return false;
    }

    UBlueprintFactory* Factory = NewObject<UBlueprintFactory>();
    Factory->ParentClass = ParentClass;

    FAssetToolsModule& AssetToolsModule = FModuleManager::LoadModuleChecked<FAssetToolsModule>("AssetTools");
    UObject* NewAsset = AssetToolsModule.Get().CreateAsset(AssetName, PackagePath, UBlueprint::StaticClass(), Factory);

    if (NewAsset)
    {
        FAssetRegistryModule::AssetCreated(NewAsset);
        NewAsset->MarkPackageDirty();
        UE_LOG(LogTemp, Log, TEXT("Wistoria Toolkit: Created Blueprint: %s"), *FullAssetPath);
        return true;
    }

    UE_LOG(LogTemp, Warning, TEXT("Wistoria Toolkit: Failed to create Blueprint: %s"), *FullAssetPath);
    return false;
}

void FWistoriaDeveloperToolkitModule::BuildCoreBlueprints()
{
    CreateFolderStructure();

    int32 CreatedOrVerified = 0;

    CreatedOrVerified += CreateBlueprintAsset(TEXT("BP_WistoriaGameMode"), TEXT("/Game/Blueprints"), TEXT("/Script/MyProject.WistoriaGameMode")) ? 1 : 0;
    CreatedOrVerified += CreateBlueprintAsset(TEXT("BP_WistoriaPlayerController"), TEXT("/Game/Blueprints"), TEXT("/Script/MyProject.WistoriaPlayerController")) ? 1 : 0;
    CreatedOrVerified += CreateBlueprintAsset(TEXT("BP_WistoriaCharacter"), TEXT("/Game/Characters"), TEXT("/Script/MyProject.WistoriaCharacter")) ? 1 : 0;
    CreatedOrVerified += CreateBlueprintAsset(TEXT("BP_WistoriaEnemy"), TEXT("/Game/Enemies"), TEXT("/Script/MyProject.WistoriaEnemy")) ? 1 : 0;
    CreatedOrVerified += CreateBlueprintAsset(TEXT("BP_DungeonBoss"), TEXT("/Game/Bosses"), TEXT("/Script/MyProject.WistoriaBossEnemy")) ? 1 : 0;
    CreatedOrVerified += CreateBlueprintAsset(TEXT("BP_DungeonZone"), TEXT("/Game/Dungeons"), TEXT("/Script/MyProject.DungeonZone")) ? 1 : 0;
    CreatedOrVerified += CreateBlueprintAsset(TEXT("BP_DungeonWaveSpawner"), TEXT("/Game/Dungeons"), TEXT("/Script/MyProject.DungeonWaveSpawner")) ? 1 : 0;
    CreatedOrVerified += CreateBlueprintAsset(TEXT("BP_SavePoint"), TEXT("/Game/Blueprints"), TEXT("/Script/MyProject.SavePoint")) ? 1 : 0;
    CreatedOrVerified += CreateBlueprintAsset(TEXT("BP_TreasureChest"), TEXT("/Game/Items"), TEXT("/Script/MyProject.TreasureChest")) ? 1 : 0;
    CreatedOrVerified += CreateBlueprintAsset(TEXT("BP_CompanionMage"), TEXT("/Game/Companions"), TEXT("/Script/MyProject.WistoriaCompanionCharacter")) ? 1 : 0;
    CreatedOrVerified += CreateBlueprintAsset(TEXT("BP_NPCProfessor"), TEXT("/Game/Blueprints"), TEXT("/Script/MyProject.DialogueNPC")) ? 1 : 0;

    UEditorLoadingAndSavingUtils::SaveDirtyPackages(true, true);

    ShowMessage(
        LOCTEXT("BlueprintsBuiltTitle", "Wistoria Toolkit"),
        FText::FromString(FString::Printf(TEXT("Core Blueprint pass complete. Created or verified %d assets."), CreatedOrVerified))
    );
}

void FWistoriaDeveloperToolkitModule::GenerateCombatTestArena()
{
    UWorld* World = GEditor ? GEditor->GetEditorWorldContext().World() : nullptr;
    if (!World)
    {
        ShowMessage(LOCTEXT("NoWorldTitle", "Wistoria Toolkit"), LOCTEXT("NoWorldMessage", "No editor world found. Open or create a level first."));
        return;
    }

    // Floor
    AStaticMeshActor* Floor = World->SpawnActor<AStaticMeshActor>(AStaticMeshActor::StaticClass(), FVector(0.f, 0.f, -50.f), FRotator::ZeroRotator);
    if (Floor)
    {
        UStaticMesh* CubeMesh = LoadObject<UStaticMesh>(nullptr, TEXT("/Engine/BasicShapes/Cube.Cube"));
        Floor->SetActorLabel(TEXT("Wistoria_TestArena_Floor"));
        Floor->SetActorScale3D(FVector(50.f, 50.f, 1.f));
        if (CubeMesh)
        {
            Floor->GetStaticMeshComponent()->SetStaticMesh(CubeMesh);
        }
        Floor->GetStaticMeshComponent()->SetCollisionProfileName(TEXT("BlockAll"));
    }

    // Player Start
    World->SpawnActor<APlayerStart>(APlayerStart::StaticClass(), FVector(0.f, -450.f, 150.f), FRotator(0.f, 0.f, 0.f));

    // Lights
    World->SpawnActor<ADirectionalLight>(ADirectionalLight::StaticClass(), FVector(0.f, 0.f, 800.f), FRotator(-45.f, 0.f, 0.f));
    World->SpawnActor<ASkyLight>(ASkyLight::StaticClass(), FVector(0.f, 0.f, 300.f), FRotator::ZeroRotator);

    // NavMesh volume
    ANavMeshBoundsVolume* NavVolume = World->SpawnActor<ANavMeshBoundsVolume>(ANavMeshBoundsVolume::StaticClass(), FVector(0.f, 0.f, 0.f), FRotator::ZeroRotator);
    if (NavVolume)
    {
        NavVolume->SetActorLabel(TEXT("Wistoria_TestArena_NavMeshBounds"));
        NavVolume->SetActorScale3D(FVector(25.f, 25.f, 5.f));
    }

    // Enemy Blueprint if present
    UBlueprint* EnemyBlueprint = LoadObject<UBlueprint>(nullptr, TEXT("/Game/Enemies/BP_WistoriaEnemy.BP_WistoriaEnemy"));
    if (EnemyBlueprint && EnemyBlueprint->GeneratedClass)
    {
        World->SpawnActor<AActor>(EnemyBlueprint->GeneratedClass, FVector(600.f, 0.f, 120.f), FRotator(0.f, 180.f, 0.f));
    }

    UEditorLoadingAndSavingUtils::SaveDirtyPackages(true, true);

    ShowMessage(
        LOCTEXT("ArenaGeneratedTitle", "Wistoria Toolkit"),
        LOCTEXT("ArenaGeneratedMessage", "Combat test arena actors were added to the current level. Run Build > Build Paths, then press P to view navigation.")
    );
}

void FWistoriaDeveloperToolkitModule::FixRedirectors()
{
    UEditorLoadingAndSavingUtils::SaveDirtyPackages(true, true);
    ShowMessage(
        LOCTEXT("RedirectorsTitle", "Wistoria Toolkit"),
        LOCTEXT("RedirectorsMessage", "Saved dirty packages. To fully fix redirectors, right-click /Game in Content Browser and choose Fix Up Redirectors in Folder.")
    );
}

#undef LOCTEXT_NAMESPACE

IMPLEMENT_MODULE(FWistoriaDeveloperToolkitModule, WistoriaDeveloperToolkit)
